File: src/gsocialsim/social/global_social_reality.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Optional, Tuple
import random

from gsocialsim.types import AgentId, TopicId
from gsocialsim.social.relationship_vector import RelationshipVector


def _clamp01(x: float) -> float:
    return 0.0 if x < 0.0 else 1.0 if x > 1.0 else x


@dataclass
class TopicReality:
    """
    Global "outside-the-agent" state for a topic.

    truth: objective ground-truth strength in [0, 1]
    salience: visibility / availability in [0, 1]
    institutional_stance: optional stance in [-1, 1] (future use)
    volatility: how fast salience decays (higher = faster)
    """
    truth: float = 0.5
    salience: float = 0.0
    institutional_stance: float = 0.0
    volatility: float = 0.02


@dataclass
class GlobalSocialReality:
    """
    Global social reality includes:
      1) Latent dyadic relationship vectors between agent pairs (symmetric).
      2) Topic-level external reality (truth + salience), which agents observe imperfectly.
    """

    # --- (1) Relationships ---
    # Key is a sorted tuple of two AgentIds to ensure R(u,v) == R(v,u)
    _relations: Dict[Tuple[AgentId, AgentId], RelationshipVector] = field(default_factory=dict)

    # --- (2) Topic reality ---
    topics: Dict[TopicId, TopicReality] = field(default_factory=dict)
    default_truth: float = 0.5
    default_volatility: float = 0.02

    # ----------------------------
    # Relationships API (unchanged)
    # ----------------------------
    def _get_key(self, u: AgentId, v: AgentId) -> Tuple[AgentId, AgentId]:
        """Creates a canonical key for a pair of agents (self-relationships allowed)."""
        if u == v:
            return (u, v)
        return tuple(sorted((u, v)))

    def get_relationship(self, u: AgentId, v: AgentId) -> RelationshipVector:
        """
        Gets the relationship between two agents.
        If no relationship exists, it creates and returns a default one.
        Self-relationships are treated as maximally trusted by default.
        """
        key = self._get_key(u, v)
        if key not in self._relations:
            if u == v:
                self._relations[key] = RelationshipVector(
                    affinity=1.0,
                    trust=1.0,
                    intimacy=1.0,
                    conflict=0.0,
                    reciprocity=1.0,
                    status_delta=0.0,
                )
            else:
                self._relations[key] = RelationshipVector()
        return self._relations[key]

    def set_relationship(self, u: AgentId, v: AgentId, vector: RelationshipVector) -> None:
        """Sets a specific relationship vector for a pair of agents."""
        key = self._get_key(u, v)
        self._relations[key] = vector

    def update_trust(self, u: AgentId, v: AgentId, delta: float) -> float:
        """
        Adjust trust for a pair, clamped to [0,1]. Returns the new trust.
        """
        rel = self.get_relationship(u, v)
        try:
            rel.trust = _clamp01(float(rel.trust) + float(delta))
        except Exception:
            rel.trust = _clamp01(rel.trust)
        return rel.trust

    # ----------------------------
    # Topic reality API (new)
    # ----------------------------
    def ensure_topic(self, topic: TopicId) -> TopicReality:
        tr = self.topics.get(topic)
        if tr is None:
            tr = TopicReality(truth=self.default_truth, salience=0.0, volatility=self.default_volatility)
            self.topics[topic] = tr
        return tr

    def truth(self, topic: TopicId) -> float:
        return self.ensure_topic(topic).truth

    def set_truth(self, topic: TopicId, value: float) -> None:
        self.ensure_topic(topic).truth = _clamp01(value)

    def salience(self, topic: TopicId) -> float:
        return self.ensure_topic(topic).salience

    def set_salience(self, topic: TopicId, value: float) -> None:
        self.ensure_topic(topic).salience = _clamp01(value)

    def bump_salience(self, topic: TopicId, delta: float) -> None:
        tr = self.ensure_topic(topic)
        tr.salience = _clamp01(tr.salience + delta)

    def decay(self) -> None:
        """Decay salience over time (call on DayBoundaryEvent, etc.)."""
        for tr in self.topics.values():
            tr.salience = _clamp01(tr.salience * (1.0 - tr.volatility))

    def observe_truth(
        self,
        topic: TopicId,
        *,
        rng: Optional[random.Random] = None,
        noise_std: float = 0.08,
        attention_gain: float = 1.0,
    ) -> float:
        """
        Noisy observation of objective truth in [0, 1].
        Higher attention_gain => less noise.
        """
        r = rng or random
        tr = self.ensure_topic(topic)
        std = max(1e-6, noise_std / max(1e-6, attention_gain))
        return _clamp01(tr.truth + r.gauss(0.0, std))

    def observe_salience(
        self,
        topic: TopicId,
        *,
        rng: Optional[random.Random] = None,
        noise_std: float = 0.05,
    ) -> float:
        r = rng or random
        tr = self.ensure_topic(topic)
        return _clamp01(tr.salience + r.gauss(0.0, noise_std))

File: cpp/src/global_social_reality.cpp
#include "global_social_reality.h"

#include <algorithm>

static inline double clamp01(double v) {
    if (v < 0.0) return 0.0;
    if (v > 1.0) return 1.0;
    return v;
}

std::pair<AgentId, AgentId> GlobalSocialReality::make_key(const AgentId& u, const AgentId& v) const {
    if (u == v) return {u, v};
    return (u < v) ? std::make_pair(u, v) : std::make_pair(v, u);
}

RelationshipVector& GlobalSocialReality::get_relationship(const AgentId& u, const AgentId& v) {
    auto key = make_key(u, v);
    auto it = relations_.find(key);
    if (it != relations_.end()) {
        return it->second;
    }
    RelationshipVector vec;
    if (u == v) {
        vec.affinity = 1.0;
        vec.trust = 1.0;
        vec.intimacy = 1.0;
        vec.conflict = 0.0;
        vec.reciprocity = 1.0;
        vec.status_delta = 0.0;
    }
    auto res = relations_.emplace(key, vec);
    return res.first->second;
}

void GlobalSocialReality::set_relationship(const AgentId& u, const AgentId& v, const RelationshipVector& vec) {
    relations_[make_key(u, v)] = vec;
}

double GlobalSocialReality::update_trust(const AgentId& u, const AgentId& v, double delta) {
    auto& rel = get_relationship(u, v);
    rel.trust = clamp01(rel.trust + delta);
    return rel.trust;
}

TopicReality& GlobalSocialReality::ensure_topic(const TopicId& topic) {
    auto it = topics_.find(topic);
    if (it != topics_.end()) return it->second;
    TopicReality tr;
    tr.truth = default_truth;
    tr.salience = 0.0;
    tr.volatility = default_volatility;
    auto res = topics_.emplace(topic, tr);
    return res.first->second;
}

double GlobalSocialReality::truth(const TopicId& topic) {
    return ensure_topic(topic).truth;
}

void GlobalSocialReality::set_truth(const TopicId& topic, double value) {
    ensure_topic(topic).truth = clamp01(value);
}

double GlobalSocialReality::salience(const TopicId& topic) {
    return ensure_topic(topic).salience;
}

void GlobalSocialReality::set_salience(const TopicId& topic, double value) {
    ensure_topic(topic).salience = clamp01(value);
}

void GlobalSocialReality::bump_salience(const TopicId& topic, double delta) {
    auto& tr = ensure_topic(topic);
    tr.salience = clamp01(tr.salience + delta);
}

void GlobalSocialReality::decay() {
    for (auto& kv : topics_) {
        kv.second.salience = clamp01(kv.second.salience * (1.0 - kv.second.volatility));
    }
}

double GlobalSocialReality::political_salience_value(const TopicId& topic) {
    return ensure_topic(topic).political_salience;
}

void GlobalSocialReality::set_political_salience(const TopicId& topic, double value) {
    ensure_topic(topic).political_salience = clamp01(value);
}

double GlobalSocialReality::observe_truth(
    const TopicId& topic,
    std::mt19937* rng,
    double noise_std,
    double attention_gain
) {
    auto& tr = ensure_topic(topic);
    double std = std::max(1e-6, noise_std / std::max(1e-6, attention_gain));
    std::normal_distribution<double> dist(0.0, std);
    if (rng) {
        return clamp01(tr.truth + dist(*rng));
    }
    std::mt19937 local_rng(std::random_device{}());
    return clamp01(tr.truth + dist(local_rng));
}

double GlobalSocialReality::observe_salience(
    const TopicId& topic,
    std::mt19937* rng,
    double noise_std
) {
    auto& tr = ensure_topic(topic);
    std::normal_distribution<double> dist(0.0, std::max(1e-6, noise_std));
    if (rng) {
        return clamp01(tr.salience + dist(*rng));
    }
    std::mt19937 local_rng(std::random_device{}());
    return clamp01(tr.salience + dist(local_rng));
}
